File: node-root/snake-ai/CalcUtils.ts
export enum ActivationFunction {
  LINEAR = "LINEAR",
  TANH = "TANH",
  RELU = "RELU",
}

export interface BaseStats {
  min: number;
  max: number;
  mean: number;
  sd: number;
  lowerQuartile: number;
  median: number;
  upperQuartile: number;
}

export default class CalcUtils {
  public static hiddenLayerActivation(x: number[], type: ActivationFunction): number[] {
    switch (type) {
      case ActivationFunction.LINEAR:
        return x;
      case ActivationFunction.TANH:
        return x.map((xx) => Math.tanh(xx));
      case ActivationFunction.RELU:
        return x.map((xx) => Math.max(0, xx));
    }
  }

  public static addition(a: number[], b: number[]) {
    return a.map((x, i) => x + b[i]);
  }

  public static multiplication(a: number[][], b: number[]) {
    return a.map((x) => x.reduce((acc, cur, idx) => acc + cur * b[idx], 0));
  }

  public static computeOneLayer(W: number[][], X: number[], B: number[]): number[] {
    return CalcUtils.addition(CalcUtils.multiplication(W, X), B);
  }

  public static computeMultipleLayer(Ws: number[][][], X: number[], Bs: number[][], hiddenLayerActivationFunction: ActivationFunction): number[] {
    const numOfLayer = Ws.length;

    let layerOutput = X;
    // hidden layer
    for (let i = 0; i < numOfLayer - 1; i++) {
      layerOutput = CalcUtils.hiddenLayerActivation(CalcUtils.computeOneLayer(Ws[i], layerOutput, Bs[i]), hiddenLayerActivationFunction);
    }

    // output layer
    layerOutput = CalcUtils.computeOneLayer(Ws[numOfLayer - 1], layerOutput, Bs[numOfLayer - 1]);
    return layerOutput;
  }

  public static indexOfMaxValueInArray(arr: number[]) {
    if (arr.length === 0) return -1;
    let max = arr[0];
    let maxIndex = 0;
    for (let i = 1; i < arr.length; i++) {
      if (arr[i] > max) {
        max = arr[i];
        maxIndex = i;
      }
    }
    return maxIndex;
  }

  public static minmax(min: number, value: number, max: number) {
    return value < min ? min : value > max ? max : value;
  }

  public static minOfArray(arr: number[]) {
    return arr.reduce((a, b) => (a < b ? a : b), arr[0]);
  }

  public static maxOfArray(arr: number[]) {
    return arr.reduce((a, b) => (a > b ? a : b), arr[0]);
  }

  public static meanOfArray(arr: number[]) {
    return arr.reduce((a, b) => a + b, 0) / arr.length;
  }

  public static sdOfArray(arr: number[]) {
    const mean = CalcUtils.meanOfArray(arr);
    return Math.sqrt(arr.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / arr.length);
  }

  public static lowerQuartileOfArray(arr: number[]) {
    const sorted = arr.slice().sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 4);
    return sorted[mid];
  }

  public static medianOfArray(arr: number[]) {
    const sorted = arr.slice().sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 2);
    return sorted[mid];
  }

  public static upperQuartileOfArray(arr: number[]) {
    const sorted = arr.slice().sort((a, b) => a - b);
    const mid = Math.floor((sorted.length * 3) / 4);
    return sorted[mid];
  }

  public static statsOfArray(arr: number[]): BaseStats {
    return {
      min: CalcUtils.minOfArray(arr),
      max: CalcUtils.maxOfArray(arr),
      mean: CalcUtils.meanOfArray(arr),
      sd: CalcUtils.sdOfArray(arr),
      lowerQuartile: CalcUtils.lowerQuartileOfArray(arr),
      median: CalcUtils.medianOfArray(arr),
      upperQuartile: CalcUtils.upperQuartileOfArray(arr),
    };
  }

  public static binarySearch(prefixSum: number[], value: number): number {
    let low = 0;
    let high = prefixSum.length - 1;
    while (low < high) {
      const mid = low + Math.floor((high - low) / 2);
      if (prefixSum[mid] >= value) {
        high = mid;
      } else {
        low = mid + 1;
      }
    }
    return low;
  }
}

File: kotlin-root/gasnake/src/main/kotlin/com/example/snake/utils/CalcUtils.kt
package com.example.snake.utils

import com.example.snake.ai.ActivationFunction
import com.example.snake.ai.BaseStats
import java.util.concurrent.ThreadLocalRandom
import kotlin.math.pow
import kotlin.math.sqrt
import kotlin.math.tanh


object CalcUtils {
    fun randomUniform(min: Double, max: Double): Double = ThreadLocalRandom.current().nextDouble(min, max)

    fun randomBool(prob: Double): Boolean = ThreadLocalRandom.current().nextDouble() < prob

    fun hiddenLayerActivation(x: DoubleArray, type: ActivationFunction): DoubleArray = when (type) {
        ActivationFunction.LINEAR -> x
        ActivationFunction.TANH -> DoubleArray(x.size) { tanh(x[it]) }
        ActivationFunction.RELU -> DoubleArray(x.size) { if (x[it] > 0) x[it] else 0.0 }
    }

    fun addition(a: DoubleArray, b: DoubleArray): DoubleArray = DoubleArray(a.size) { a[it] + b[it] }

    fun multiplication(a: List<DoubleArray>, b: DoubleArray): DoubleArray = DoubleArray(a.size) { i ->
        var sum = 0.0
        for (j in a[i].indices) {
            sum += a[i][j] * b[j]
        }
        sum
    }

    fun computeOneLayer(W: List<DoubleArray>, X: DoubleArray, B: DoubleArray): DoubleArray =
        addition(multiplication(W, X), B)

    fun computeMultipleLayer(
        Ws: List<List<DoubleArray>>,
        X: DoubleArray,
        Bs: List<DoubleArray>,
        hiddenLayerActivationFunction: ActivationFunction
    ): DoubleArray {
        var layerOutput = X
        // hidden layer
        for (i in 0 until Ws.size - 1) {
            layerOutput =
                hiddenLayerActivation(computeOneLayer(Ws[i], layerOutput, Bs[i]), hiddenLayerActivationFunction)
        }

        // output layer
        layerOutput = computeOneLayer(Ws.last(), layerOutput, Bs.last())
        return layerOutput
    }

    fun indexOfMaxValueInArray(arr: DoubleArray): Int = arr.indices.maxByOrNull { arr[it] } ?: -1

    fun minmax(min: Double, value: Double, max: Double): Double {
        return when {
            value < min -> min
            value > max -> max
            else -> value
        }
    }

    fun minOfArray(arr: DoubleArray): Double = arr.minOrNull() ?: throw IllegalArgumentException("Array is empty")

    fun maxOfArray(arr: DoubleArray): Double = arr.maxOrNull() ?: throw IllegalArgumentException("Array is empty")

    fun meanOfArray(arr: DoubleArray): Double = arr.average()

    fun sdOfArray(arr: DoubleArray): Double {
        val mean = meanOfArray(arr)
        return sqrt(arr.map { (it - mean).pow(2) }.average())
    }

    fun lowerQuartileOfArray(arr: DoubleArray): Double {
        val sorted = arr.sorted()
        val mid = sorted.size / 4
        return sorted[mid]
    }

    fun medianOfArray(arr: DoubleArray): Double {
        val sorted = arr.sorted()
        val mid = sorted.size / 2
        return sorted[mid]
    }

    fun upperQuartileOfArray(arr: DoubleArray): Double {
        val sorted = arr.sorted()
        val mid = (sorted.size * 3) / 4
        return sorted[mid]
    }

    fun statsOfArray(arr: DoubleArray): BaseStats = BaseStats(
        min = minOfArray(arr),
        max = maxOfArray(arr),
        mean = meanOfArray(arr),
        sd = sdOfArray(arr),
        lowerQuartile = lowerQuartileOfArray(arr),
        median = medianOfArray(arr),
        upperQuartile = upperQuartileOfArray(arr)
    )

    fun binarySearch(prefixSum: List<Double>, value: Double): Int {
        var low = 0
        var high = prefixSum.size - 1
        while (low < high) {
            val mid = low + (high - low) / 2
            if (prefixSum[mid] >= value) high = mid
            else low = mid + 1
        }
        return low
    }

}